File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/maintain/AuthUserGroupService.java
package uk.gov.justice.digital.hmpps.oauth2server.maintain;

import com.microsoft.applicationinsights.TelemetryClient;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.hibernate.Hibernate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.Group;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.GroupRepository;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserEmailRepository;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;


@Service
@Slf4j
@AllArgsConstructor
@Transactional(transactionManager = "authTransactionManager", readOnly = true)
public class AuthUserGroupService {
    private final UserEmailRepository userEmailRepository;
    private final GroupRepository groupRepository;
    private final TelemetryClient telemetryClient;

    @Transactional(transactionManager = "authTransactionManager")
    public void addGroup(final String username, final String groupCode, final String modifier) throws AuthUserGroupException {
        // already checked that user exists
        final var userEmail = userEmailRepository.findByUsernameAndMasterIsTrue(username).orElseThrow();

        final var groupFormatted = formatGroup(groupCode);

        // check that group exists
        final var group = groupRepository.findByGroupCode(groupFormatted).orElseThrow(() -> new AuthUserGroupException("group", "notfound"));

        if (userEmail.getGroups().contains(group)) {
            throw new AuthUserGroupExistsException();
        }

        // TODO: Validate that the group is allowed to be added to the user:
        // 1. If super user then can add anyone to anything
        // 2. If group admin then needs to be one of their groups and user can't be a member of a different group

        log.info("Adding group {} to user {}", groupFormatted, username);
        userEmail.getGroups().add(group);
        telemetryClient.trackEvent("AuthUserGroupAddSuccess", Map.of("username", username, "group", groupFormatted, "admin", modifier), null);
        userEmailRepository.save(userEmail);
    }

    @Transactional(transactionManager = "authTransactionManager")
    public void removeGroup(final String username, final String groupCode, final String modifier) throws AuthUserGroupException {
        final var groupFormatted = formatGroup(groupCode);
        // already checked that user exists
        final var userEmail = userEmailRepository.findByUsernameAndMasterIsTrue(username).orElseThrow();

        if (userEmail.getGroups().stream().map(Group::getGroupCode).noneMatch(a -> a.equals(groupFormatted))) {
            throw new AuthUserGroupException("group", "missing");
        }

        log.info("Removing group {} from user {}", groupFormatted, username);
        userEmail.getGroups().removeIf(a -> a.getGroupCode().equals(groupFormatted));

        telemetryClient.trackEvent("AuthUserGroupRemoveSuccess", Map.of("username", username, "group", groupCode, "admin", modifier), null);
        userEmailRepository.save(userEmail);
    }

    private String formatGroup(final String group) {
        return StringUtils.upperCase(StringUtils.trim(group));
    }

    public List<Group> getAllGroups() {
        return groupRepository.findAll();
    }

    public Optional<Set<Group>> getAuthUserGroups(final String username) {
        final var user = userEmailRepository.findByUsernameAndMasterIsTrue(StringUtils.upperCase(StringUtils.trim(username)));
        return user.map(u -> {
            Hibernate.initialize(u.getGroups());
            return u.getGroups();
        });
    }

    public static class AuthUserGroupExistsException extends AuthUserGroupException {
        public AuthUserGroupExistsException() {
            super("group", "exists");
        }
    }


    @Getter
    public static class AuthUserGroupException extends Exception {
        private final String errorCode;
        private final String field;

        public AuthUserGroupException(final String field, final String errorCode) {
            super(String.format("Add group failed for field %s with reason: %s", field, errorCode));

            this.field = field;
            this.errorCode = errorCode;
        }
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/maintain/AuthUserGroupService.kt
package uk.gov.justice.digital.hmpps.oauth2server.maintain

import com.microsoft.applicationinsights.TelemetryClient
import org.hibernate.Hibernate
import org.slf4j.LoggerFactory
import org.springframework.data.repository.findByIdOrNull
import org.springframework.security.core.GrantedAuthority
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.Group
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.GroupRepository
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository
import uk.gov.justice.digital.hmpps.oauth2server.security.MaintainUserCheck
import uk.gov.justice.digital.hmpps.oauth2server.security.MaintainUserCheck.AuthUserGroupRelationshipException
import uk.gov.justice.digital.hmpps.oauth2server.security.MaintainUserCheck.Companion.canMaintainAuthUsers
import java.util.UUID

@Service
@Transactional(transactionManager = "authTransactionManager", readOnly = true)
class AuthUserGroupService(
  private val userRepository: UserRepository,
  private val groupRepository: GroupRepository,
  private val maintainUserCheck: MaintainUserCheck,
  private val telemetryClient: TelemetryClient
) {

  companion object {
    private val log = LoggerFactory.getLogger(this::class.java)
  }

  @Transactional(transactionManager = "authTransactionManager")
  @Throws(
    AuthUserGroupException::class,
    AuthUserGroupManagerException::class,
    AuthUserGroupRelationshipException::class
  )
  fun addGroup(username: String, groupCode: String, modifier: String, authorities: Collection<GrantedAuthority>) {
    // already checked that user exists
    val user = userRepository.findByUsernameAndMasterIsTrue(username).orElseThrow()
    val groupFormatted = formatGroup(groupCode)

    // check that group exists
    val group =
      groupRepository.findByGroupCode(groupFormatted) ?: throw AuthUserGroupException("group", "notfound")
    if (user.groups.contains(group)) {
      throw AuthUserGroupExistsException()
    }
    // check that modifier is able to add user to group
    if (!checkGroupModifier(groupCode, authorities, modifier)) {
      throw AuthUserGroupManagerException("Add", "group", "managerNotMember")
    }
    // check that modifier is able to maintain the user
    maintainUserCheck.ensureUserLoggedInUserRelationship(modifier, authorities, user)

    log.info("Adding group {} to user {}", groupFormatted, username)
    user.groups.add(group)
    user.authorities.addAll(group.assignableRoles.filter { it.automatic }.map { it.role })
    telemetryClient.trackEvent(
      "AuthUserGroupAddSuccess",
      mapOf("username" to username, "group" to groupFormatted, "admin" to modifier),
      null
    )
  }

  @Transactional(transactionManager = "authTransactionManager")
  @Throws(AuthUserGroupException::class, AuthUserGroupManagerException::class, AuthUserLastGroupException::class)
  fun removeGroup(username: String, groupCode: String, modifier: String, authorities: Collection<GrantedAuthority>) {
    val groupFormatted = formatGroup(groupCode)
    // already checked that user exists
    val user = userRepository.findByUsernameAndMasterIsTrue(username).orElseThrow()
    if (user.groups.map { it.groupCode }.none { it == groupFormatted }
    ) {
      throw AuthUserGroupException("group", "missing")
    }
    if (!checkGroupModifier(groupFormatted, authorities, modifier)) {
      throw AuthUserGroupManagerException("delete", "group", "managerNotMember")
    }

    if (user.groups.count() == 1 && !canMaintainAuthUsers(authorities)) {
      throw AuthUserLastGroupException("group", "last")
    }

    log.info("Removing group {} from user {}", groupFormatted, username)
    user.groups.removeIf { a: Group -> a.groupCode == groupFormatted }
    telemetryClient.trackEvent(
      "AuthUserGroupRemoveSuccess",
      mapOf("username" to username, "group" to groupCode, "admin" to modifier),
      null
    )
  }

  private fun checkGroupModifier(
    groupCode: String,
    authorities: Collection<GrantedAuthority>,
    modifier: String
  ): Boolean {
    return if (canMaintainAuthUsers(authorities)) {
      true
    } else {
      val modifierGroups = getAssignableGroups(modifier, authorities)
      return modifierGroups.map { it.groupCode }.contains(groupCode)
    }
  }

  private fun formatGroup(group: String) = group.trim().uppercase()

  val allGroups: List<Group>
    get() = groupRepository.findAllByOrderByGroupName()

  fun getAuthUserGroups(username: String): Set<Group>? {
    val user = userRepository.findByUsernameAndMasterIsTrue(username.trim().uppercase())
    return user.map { u: User ->
      Hibernate.initialize(u.groups)
      u.groups.forEach { Hibernate.initialize(it.children) }
      u.groups
    }.orElse(null)
  }

  fun getAuthUserGroupsByUserId(userId: String, admin: String, authorities: Collection<GrantedAuthority>): Set<Group>? =
    userRepository.findByIdOrNull(UUID.fromString(userId))?.let { u: User ->
      maintainUserCheck.ensureUserLoggedInUserRelationship(admin, authorities, u)
      Hibernate.initialize(u.groups)
      u.groups.forEach { Hibernate.initialize(it.children) }
      u.groups
    }

  fun getAssignableGroups(username: String, authorities: Collection<GrantedAuthority>): List<Group> =
    if (canMaintainAuthUsers(authorities)) allGroups.toList()
    else getAuthUserGroups(username)?.sortedBy { it.groupName } ?: listOf()

  class AuthUserGroupExistsException : AuthUserGroupException("group", "exists")
  open class AuthUserGroupException(val field: String, val errorCode: String) :
    Exception("Add group failed for field $field with reason: $errorCode")

  open class AuthUserGroupManagerException(val action: String = "add", val field: String, val errorCode: String) :
    Exception("$action group failed for field $field with reason: $errorCode")

  open class AuthUserLastGroupException(val field: String, val errorCode: String) :
    Exception("remove group failed for field $field with reason: $errorCode")
}
